lib/src/client/transport/xhr_transport.dart

```diff
@@ -0,0 +1,152 @@
+
+  @override
+  StreamSink<List<int>> get outgoingMessages => _outgoingMessages.sink;
+
+  XhrTransportStream(this._request) {
+    _incomingProcessor = new StreamController();
+    _incomingMessages = new StreamController();
+    _outgoingMessages = new StreamController();
+
+    _incomingProcessor.stream
+        .transform(new GrpcWebDecoder())
+        .transform(grpcDecompressor())
+        .listen(_incomingMessages.add,
+            onError: _incomingMessages.addError,
+            onDone: _incomingMessages.close);
+
+    _outgoingMessages.stream
+        .map(GrpcHttpEncoder.frame)
+        .listen((data) => _request.send(data));
+
+    _request.onReadyStateChange.listen((data) {
+      final contentType = _request.getResponseHeader('Content-Type');
+      if (contentType == null) {
+        return;
+      }
+
+      if (_request.readyState == HttpRequest.HEADERS_RECEIVED) {
+        if (contentType.startsWith('application/grpc')) {
+          if (_request.response == null) {
+            return;
+          }
+
+          // Force a metadata message with headers
+          final headers = new GrpcMetadata(_request.responseHeaders);
+          _incomingMessages.add(headers);
+        }
+      }
+
+      if (_request.readyState == HttpRequest.DONE) {
+        _incomingProcessor.close();
+        _outgoingMessages.close();
+      }
+    });
+
+    _request.onProgress.listen((data) {
+      final responeString = _request.response as String;
+      final rawText = (responeString).substring(_requestBytesRead);
+      _requestBytesRead = responeString.length;
+      final bytes = _stringToArrayBuffer(rawText);
+      _incomingProcessor.add(bytes);
+    });
+  }
+
+  @override
+  Future<void> terminate() async {
+    await _incomingProcessor.close();
+    await _outgoingMessages.close();
+    _request.abort();
+  }
+
+  ByteBuffer _stringToArrayBuffer(String str) {
+    final codePoints = new Uint8List(str.length);
+    var arrayIndex = 0;
+    for (var i = 0; i < str.length; i++) {
+      final codePoint = str.codeUnitAt(i);
+      codePoints[arrayIndex++] = codePoint & 0xFF;
+    }
+    return codePoints.buffer;
+  }
+}
+
+class XhrTransport extends Transport {
+  final String host;
+  final int port;
+  final ChannelOptions options;
+
+  HttpRequest _request;
+
+  XhrTransport(this.host, this.port, this.options);
+
+  @override
+  Future<void> connect() async {}
+
+  @override
+  Future<void> finish() async {
+    // TODO: implement finish
+  }
+
+  @visibleForTesting
+  void initializeRequest(HttpRequest request, Map<String, String> metadata) {
+    for (final header in metadata.keys) {
+      request.setRequestHeader(header, metadata[header]);
+    }
+    request.setRequestHeader('Content-Type', 'application/grpc-web+proto');
+    request.setRequestHeader('X-User-Agent', 'grpc-web-dart/0.1');
+    request.setRequestHeader('X-Grpc-Web', '1');
+    request.overrideMimeType('text/plain; charset=x-user-defined');
```

Comment:
Yes -- this and the use of `_request.response as String` from above allow me to process the request streaming instead of processing it after it closes. This is to allow Server Side Streaming requests.

Reply:
👍 Done!